### PR TITLE
feat: Space3 Elite contract

### PR DESCRIPTION
**Pull request by sangtran-a8**

### CODE REVIEW

File: contracts/nfts/interfaces/IDelayedReveal.sol [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+interface IDelayedReveal {
+  /// @dev Emitted when tokens are revealed.
```
Comment (thanhle-ancient8) [inline contracts/nfts/interfaces/IDelayedReveal.sol:5]:
```suggestion
  /* ================ Events ================ */
  /// @dev Emitted when tokens are revealed.
```

File: contracts/nfts/interfaces/IDelayedReveal.sol [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  event TokenURIRevealed(uint256 indexed index, string revealedURI);
+
+  /// @dev The contract doesn't have any URL to be delayed revealed
```
Comment (thanhle-ancient8) [inline contracts/nfts/interfaces/IDelayedReveal.sol:8]:
```suggestion
  /* ================ Errors ================ */
  /// @dev The contract doesn't have any URL to be delayed revealed
```

File: contracts/nfts/interfaces/IDelayedReveal.sol [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  error DelayedRevealNothingToReveal();
+
+  /// @dev The result of the returned incorrect hash
+  error DelayedRevealIncorrectResultHash(bytes32 expected, bytes32 actual);
+
+  /**
```
Comment (thanhle-ancient8) [inline contracts/nfts/interfaces/IDelayedReveal.sol:14]:
```suggestion
  /* ================ Functions ================ */
  /**
```

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import '@openzeppelin/contracts/utils/Multicall.sol';
+import '../nfts/DelayedReveal.sol';
+
+contract Space3EliteSoul is
+  ERC721URIStorage,
+  AccessControl,
+  Multicall,
+  DelayedReveal
+{
+  uint256 private _nextTokenId;
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:15]:
```suggestion
  uint256 public nextTokenId;
```

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  bytes32 public constant MINTER_ROLE = keccak256('MINTER_ROLE');
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:16]: constant should be placed upper stage values

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  /* ============= Modifiers ============= */
+
+  modifier onlyOwnerOrAdmin(uint256 tokenId) {
+    require(_canReveal(tokenId), 'Not authorized!');
+    _;
+  }
+
+  /* ============= Constructor ============= */
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:25]:
```suggestion
```

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  {
+    return
+      ERC721URIStorage.supportsInterface(interfaceId) ||
+      AccessControl.supportsInterface(interfaceId) ||
+      DelayedReveal.supportsInterface(interfaceId);
+  }
+
+  function nextTokenId() public view returns (uint256) {
+    return _nextTokenId;
+  }
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:90]:
```suggestion
```

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    bytes calldata key
+  )
+    external
+    override
+    onlyOwnerOrAdmin(tokenId)
+    returns (string memory revealedURI)
+  {
+    revealedURI = super._reveal(tokenId, key);
+    _setTokenURI(tokenId, revealedURI);
+    return revealedURI;
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:69]: missing emit event?

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  AccessControl,
+  Multicall,
+  DelayedReveal
+{
+  uint256 private _nextTokenId;
+  bytes32 public constant MINTER_ROLE = keccak256('MINTER_ROLE');
+
+  /* ============= Modifiers ============= */
+
+  modifier onlyOwnerOrAdmin(uint256 tokenId) {
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:20]:
```suggestion
  modifier canReveal(uint256 tokenId) {
```

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    require(_canReveal(tokenId), 'Not authorized!');
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:21]: this hiding error

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  /* ============= Internal Functions ============= */
+
+  /// @dev Checks whether NFT can be revealed in the given execution context.
+  function _canReveal(uint256 tokenId) internal view virtual returns (bool) {
+    return
+      hasRole(DEFAULT_ADMIN_ROLE, msg.sender) ||
+      msg.sender == _ownerOf(tokenId);
+  }
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:99]:
```suggestion
  function _checkRevealAuthority(uint256 tokenId) internal view virtual {
    require(
      hasRole(DEFAULT_ADMIN_ROLE, msg.sender) ||
      msg.sender == _ownerOf(tokenId),
      "Only token owner or admin can reveal"
      );
  }
```

File: deploy/regen-721-engine/01_deploy_Space3EliteSoul.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+const func: DeployFunction = async function ({
+  getNamedAccounts,
+  deployments,
+  run,
+}: HardhatRuntimeEnvironment) {
+  const { deployer } = await getNamedAccounts()
+  const { deploy } = deployments
+  const tokenName = process.env.SPACE3_ELITE_TOKEN_NAME || ''
+  const tokenSymbol = process.env.SPACE3_ELITE_TOKEN_SYMBOL || ''
```
Comment (thanhle-ancient8) [inline deploy/regen-721-engine/01_deploy_Space3EliteSoul.ts:12]:
```suggestion
```

File: deploy/regen-721-engine/01_deploy_Space3EliteSoul.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+  const { address } = await deploy('Space3EliteSoul', {
+    from: deployer,
+    args: [deployer, tokenName, tokenSymbol],
```
Comment (thanhle-ancient8) [inline deploy/regen-721-engine/01_deploy_Space3EliteSoul.ts:16]:
```suggestion
    args: [deployer],
```

File: test/regen-721-engine/Space3EliteSoul.spec.ts [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+    return { owner, wallet0, others, Space3EliteSoulNFT }
+  }
+
+  describe('#Deployment', function () {
+    it('Should deploy contract successfully', async function () {
+      const { Space3EliteSoulNFT } = await loadFixture(deployFixture)
+
+      expect(await Space3EliteSoulNFT.getAddress()).to.be.match(
+        /^0x[a-fA-F0-9]{40}$/,
```
Comment (thanhle-ancient8) [inline test/regen-721-engine/Space3EliteSoul.spec.ts:37]: overkill, You think what kind of address will be, Solana??

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    require(_canReveal(tokenId), 'Not authorized!');
+    _;
+  }
+
+  /* ============= Constructor ============= */
+
+  constructor(
+    address defaultAdmin,
+    string memory name,
+    string memory symbol
```
[the remark this answers is not in the document]
Comment (sangtran-a8) [reply contracts/regen-721-engine/Space3EliteSoul.sol:30]: collecting...

File: contracts/regen-721-engine/Space3EliteSoul.sol (hunk 8 above)
Comment (sangtran-a8) [reply contracts/regen-721-engine/Space3EliteSoul.sol:69]: already have an event `MetadataUpdate` in `_setTokenURI`
Do we need a new event here?

File: test/regen-721-engine/Space3EliteSoul.spec.ts (hunk 14 above)
Comment (sangtran-a8) [reply test/regen-721-engine/Space3EliteSoul.spec.ts:37]: but still ok??? What do u want to change to?

File: contracts/regen-721-engine/Space3EliteSoul.sol (hunk 8 above)
Comment (thanhle-ancient8) [reply contracts/regen-721-engine/Space3EliteSoul.sol:69]: this is reveal event

Comment (sangtran-a8) [reply contracts/regen-721-engine/Space3EliteSoul.sol:69]: ok luôn

Comment (sangtran-a8) [reply contracts/regen-721-engine/Space3EliteSoul.sol:69]: wait, I have double-checked, revealed event has been emitted in `super._reveal` function

**Comment by sangtran-a8** [comment]
Screenshot 2024-08-06 at 18 23 35

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        msg.sender == _ownerOf(tokenId),
+      'Only token owner or admin can reveal!'
+    );
+    _;
+  }
+
+  constructor(address defaultAdmin) ERC721('Space3 Elite Soul NFT', 'SEN') {
+    _grantRole(DEFAULT_ADMIN_ROLE, defaultAdmin);
+    _grantRole(MINTER_ROLE, defaultAdmin);
+  }
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:32]:
```suggestion
  constructor() ERC721('Space3 Elite Soul NFT', 'SEN') {
    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
    _grantRole(MINTER_ROLE, msg.sender);
  }
```

File: contracts/regen-721-engine/Space3EliteSoul.sol [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  function mint(
+    address to,
+    bytes memory _encryptedData
+  ) external onlyRole(MINTER_ROLE) {
+    uint256 tokenId = nextTokenId++;
+    _safeMint(to, tokenId);
+    _setEncryptedData(tokenId, _encryptedData);
+  }
+
```
Comment (thanhle-ancient8) [inline contracts/regen-721-engine/Space3EliteSoul.sol:49]: Missing `bulkMint(address[] tos, bytes[] memory _encryptedData)`

File: example.env [hunk 18]
```diff
@@ -12,3 +12,7 @@ A8_CHAIN_ID=28122024
 
 # Regen 721 Engine
 SPACE3_GENESIS_TOKEN_BASE_URI=https://space3-inventory.s3.ap-southeast-1.amazonaws.com/space3-genesis-testnet/
+
+# Space3 Elite Soul NFT
+SPACE3_ELITE_TOKEN_NAME=
+SPACE3_ELITE_TOKEN_SYMBOL=
```
Comment (thanhle-ancient8) [inline example.env:18]:
```suggestion
```

**Pull request merged**